File: data-structures/linked-list/java/linked_list_OOP.java
/*
Raspberry Pi Foundation
Developed as part of Ada Computer Science
Usage licensed under CC BY-NC-SA 4.0

Note: This file is designed to be copied out and compiled on your machine.
In order for the program to compile properly in some IDEs, you need to ensure that the
class filename is the same as the class name with the main method in it.

To run this file you need to:
1. Copy the contents
2. Create a Java project in the IDE of your choice (Eclipse, for example)
3. Create a new Java Class that uses the same name as the class containing the main method in this program
4. Paste the contents into the new Java Class
5. Save and run the program
*/

// A class for a node in the linked list
class Node {
    private String data;
    private Node nextItem;

    // Constructor method
    public Node(String givenData) {
        data = givenData;
    }

    public String getData() {
        return data;
    }

    public Node getNext() {
        return nextItem;
    }

    public void setNext(Node newNext) {
        nextItem = newNext;
    }
}


// A class for the linked list
class LinkedList {
    private Node head; // Do not initialise yet as the linked list is empty

    public Node getHead() {
        return head;
    }

    public void setHead(Node newHead) {
        head = newHead;
    }

    // Insert a node to the front of the list
    public void insertAtFront(String data) {
        // Create a new node
        Node newNode = new Node(data);

        // Check if the head node exists
        if (head == null) {
            setHead(newNode);
        }
        else {
            // Update the pointers so the new node is the head
            newNode.setNext(head);
            setHead(newNode);
        }
    }


    // Insert a node into the correct position in an ordered list
    public void insertInOrder(String data) {
        // Create a new node
        Node newNode = new Node(data);

        // Start at the head of the list
        Node current = getHead();

        // Check if there are no nodes in the list
        if (current == null) {
            setHead(newNode);
        }

        // Check if the new node data is before the head data
        else if (newNode.getData().compareTo(current.getData()) < 0) {
            // Set the new node as the head of the list
            newNode.setNext(head);
            setHead(newNode);
        }

        // Otherwise find where the new node should be positioned
        else {
            // Repeat until the point of insertion is found
            while (current.getNext() != null
                    && current.getNext().getData().compareTo(newNode.getData()) < 0) {
                // Get the next node
                current = current.getNext();
            }
            // Update the pointers of the new and current nodes
            newNode.setNext(current.getNext());
            current.setNext(newNode);
        }
    }


    // Traverse the list and output the data from each node
    public void traverse() {
        // Set the current node as the head
        Node current = getHead();

        // Repeat until there are no more linked nodes
        while (current != null) {
            System.out.println(current.getData());
            current = current.getNext();
        }
    }


    // Delete a node. This assumes that the node does exist in the list
    public void delete(String data) {
        // Start at the head of the list
        Node current = getHead();

        // Check if the head node is to be deleted
        if (current.getData().equals(data)) {
            // Update the head pointer
            setHead(current.getNext());
        }
        else {
            // Repeat until the node has been found
            while (!current.getNext().getData().equals(data)) {
                // Change the current node to be the next node
                current = current.getNext();
            }
            // Set the pointer to be the next node's pointer
            current.setNext(current.getNext().getNext());
        }
    }


}


class DataStructures {
    // The main method is the entry point for all Java programs
    public static void main(String[] args) {
        System.out.println("### Linked list (OOP) ###");

        // Instantiate an empty linked list object
        LinkedList myList = new LinkedList();

        // Insert test data into the linked list
        System.out.println("\nInsert test data into the linked list in order:");
        insertTestData(myList);
        myList.traverse();

        // Testing - insert at front
        System.out.println("\nInsert a node to the front of the list:");
        myList.insertAtFront("Albus");
        myList.traverse();

        // Testing - insert in order
        System.out.println("\nInsert a node to the end of the list:");
        myList.insertInOrder("Zeb");
        myList.traverse();

        // Testing - delete a node
        System.out.println("\nDelete a node from the list:");
        myList.delete("Sabrina");
        myList.traverse();
    }


    // Insert test data into the linked list
    public static void insertTestData(LinkedList myList) {
        myList.insertInOrder("Julie");
        myList.insertInOrder("Rey");
        myList.insertInOrder("Habib");
        myList.insertInOrder("Sabrina");
        myList.insertInOrder("Mina");
    }


}

File: data-structures/linked-list/visual-basic/linked_list_OOP.vb
' Raspberry Pi Foundation
' Developed as part of Ada Computer Science
' Usage licensed under CC BY-NC-SA 4.0
' Note: This file is designed to be copied out and compiled on your machine.
' To run this file you need to:
' 1. Copy the contents
' 2. Paste them into the Visual Basic IDE of your choice (Visual Studio, for example)
' 3. Compile the program
' 4. Run the program


' A class to represent a node in a linked list
Class Node
    Private data As String
    Private nextItem As Node

    Public Sub New(ByVal givenData As String)
        data = givenData
    End Sub

    Public Function GetData() As String
        Return data
    End Function

    Public Function GetNext() As Node
        Return nextItem
    End Function

    Public Sub SetNext(ByVal newNext As Node)
        nextItem = newNext
    End Sub
End Class


' A class to represent a linked list
Class LinkedList
    Private head As Node  ' Do not intialise yet as the linked list is empty


    ' Insert a node to the front of the list
    Public Sub InsertAtFront(ByVal data As String)

        ' Create a new node
        Dim newNode As Node = New Node(data)

        ' Check if the head node exists
        If head Is Nothing Then
            head = newNode
        Else
            ' Update the pointers so the new node is the head
            newNode.SetNext(head)
            head = newNode
        End If
    End Sub


    ' Insert a node into the correct position in an ordered list
    Public Sub InsertInOrder(ByVal data As String)

        ' Create a new node
        Dim newNode As Node = New Node(data)

        ' Start at the head of the list
        Dim current As Node = head

        ' Check if there are no nodes in the list
        If current Is Nothing Then
            head = newNode

            ' Check if the new node data is before the head data
        ElseIf String.Compare(newNode.GetData(), current.GetData()) < 0 Then
            ' Set the new node as the head of the list
            newNode.SetNext(head)
            head = newNode
            ' Otherwise find where the new node should be positioned
        Else
            ' Repeat until the point of insertion is found
            While current.GetNext() IsNot Nothing AndAlso String.Compare(current.GetNext().GetData(), newNode.GetData()) < 0
                ' Get the next node
                current = current.GetNext()
            End While
            ' Update the pointers of the new and current nodes
            newNode.SetNext(current.GetNext())
            current.SetNext(newNode)
        End If
    End Sub


    ' Traverse the list and output the data from each node
    Public Sub Traverse()
        ' Set the current node as the head
        Dim current As Node = head

        ' Repeat until there are no more linked nodes
        While current IsNot Nothing
            Console.WriteLine(current.GetData())
            current = current.GetNext()
        End While
    End Sub

    ' Delete a node. This assumes that the node does exist in the list
    Public Sub Delete(ByVal data As String)
        ' Start at the head of the list
        Dim current As Node = head

        ' Check if the current node is the node to be deleted
        If current.GetData() = data Then
            ' Update the head pointer
            head = current.GetNext()
        Else
            ' Repeat until the node has been found
            While current.GetNext().GetData() <> data
                ' Change the current node to be the next node
                current = current.GetNext()
            End While
            ' Set the pointer to be the next node's pointer
            current.SetNext(current.GetNext().GetNext())
        End If
    End Sub


End Class

Module Program
    ' The Main subroutine is the default entry point for a VB program
    Sub Main()
        Console.WriteLine("### Linked list (OOP) ###")

        ' Instantiate an empty linked list object
        Dim myList As LinkedList = New LinkedList()

        ' Insert test data into the linked list
        Console.WriteLine(vbLf & "Insert test data into the linked list in order:")
        InsertTestData(myList)
        myList.Traverse()

        ' Testing - insert at front
        Console.WriteLine(vbLf & "Insert a node to the front of the list:")
        myList.InsertAtFront("Albus")
        myList.Traverse()

        ' Testing - insert at end
        Console.WriteLine(vbLf & "Insert a node to the end of the list:")
        myList.InsertInOrder("Zeb")
        myList.Traverse()

        ' Testing - delete a node
        Console.WriteLine(vbLf & "Delete a node from the list:")
        myList.Delete("Sabrina")
        myList.Traverse()
    End Sub


    ' Insert test data into the linked list
    Sub InsertTestData(ByVal myList As LinkedList)
        myList.InsertInOrder("Julie")
        myList.InsertInOrder("Rey")
        myList.InsertInOrder("Habib")
        myList.InsertInOrder("Sabrina")
        myList.InsertInOrder("Mina")
    End Sub


End Module
